segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,308 @@
+    }
+
+    private EventsReadAndTruncationPoints readEventsFromSegment(EventProcessorImpl eventProcessor, int currentOffset) {
+        List<BufferView> readEvents = new ArrayList<>();
+        int finalTruncationOffset = currentOffset;
+        while (readEvents.size() < eventProcessor.getConfig().getMaxItemsAtOnce()) {
+            try {
+                int appendLength = bufferViewToInt(readAppendOfLength(eventProcessor.segment, finalTruncationOffset, Integer.BYTES));
+                finalTruncationOffset += Integer.BYTES;
+                BufferView data = readAppendOfLength(eventProcessor.segment, finalTruncationOffset, appendLength);
+                readEvents.add(data);
+                finalTruncationOffset += data.getLength();
+            } catch (Exception e) {
+                // We got some data, do not throw, just stop looping.
+                if (!readEvents.isEmpty()) {
+                    break;
+                }
+                // TimeoutException is expected if there is no data in the Segment. Otherwise it may be a problem.
+                throw (e instanceof TimeoutException) ? new NoDataAvailableException() : new CompletionException(e);
+            }
+        }
+        return new EventsReadAndTruncationPoints(readEvents, currentOffset, finalTruncationOffset);
+    }
+
+    private CompletableFuture<Void> updateAndTruncateInternalSegment(EventProcessorImpl ep,
+                                                                     EventsReadAndTruncationPoints readResult,
+                                                                     Timer iterationTime) {
+        return updateLastProcessedEventOffset(ep.segment, readResult.getFinalTruncationOffset())
+                .thenCompose(v -> ep.segment.truncate(readResult.getFinalTruncationOffset(), OPERATION_TIMEOUT))
+                .thenAccept(v -> {
+                    // Decrement outstanding bytes and publish the last latency value after successful truncation.
+                    ep.outstandingBytes.addAndGet(-readResult.getReadBytes());
+                    ep.lastIterationLatency.set(iterationTime.getElapsedMillis());
+                });
+    }
+
+    private BufferView readAppendOfLength(DirectSegmentAccess segment, int offset, int length) throws Exception {
+        @Cleanup
+        ReadResult readResult = segment.read(offset, length, OPERATION_TIMEOUT);
+        ReadResultEntry entry = readResult.next();
+        entry.requestContent(OPERATION_TIMEOUT);
+        return entry.getContent().get(OPERATION_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    private int bufferViewToInt(BufferView appendLengthContent) throws IOException {
+        return ByteBuffer.wrap(appendLengthContent.getReader().readAllBytes()).getInt();
+    }
+
+    private ByteArraySegment intToByteArraySegment(int intValue) {
+        return new ByteArraySegment(ByteBuffer.allocate(Integer.BYTES).putInt(intValue).array());
```

Comment:
```
val r = new ByteArraySegment(Integer.BYTES);
r.setInt(0, intValue);
return r;
```

That should do it.

Also how are you handling future versions? Consider that in the future we may want to update the format for this. How can you do that? One option is to reserve the first 1 byte of the Int for Version, and use the remaining 3 for the length (that gives you a max of 24MB - more than needed).

Reply:
Thanks, based on this comment, I have added a new class called `ContainerEventProcessor.ProcessorEventSerializer` that helps to (de)serialize `ProcessorEventData` objects. These objects contain: i) version (byte), ii) length (int), iii) and data (BufferView). Also, we set a max total size (including header) for the individual events written to an `EventProcessor` (1MB).